File: archive-py/skills_usage/hook_install.py
"""Install/uninstall the Skill PreToolUse + PostToolUse hooks in ~/.claude/settings.json.

Idempotent. Each operation:
1. Backs up the current settings.json with a timestamped suffix (and updates
   ~/.claude/settings.json.pre-skills-usage.bak if missing).
2. Edits the JSON in-memory.
3. Atomically replaces settings.json via a temp-file + os.replace.

Honors --scope user|project|local. The CLI passes scope=user by default.
"""
from __future__ import annotations

import json
import os
import shutil
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

HOOK_SCRIPT = Path(__file__).resolve().parent.parent / "hooks" / "skill_event.py"
PRISTINE_BACKUP = Path.home() / ".claude" / "settings.json.pre-skills-usage.bak"

# Marker so we can recognize and remove our own hooks unambiguously.
HOOK_MARKER = "skills-usage:skill-events"


@dataclass
class Scope:
    name: str       # "user" | "project" | "local"
    path: Path
    label: str


def resolve_scope(scope: str, project_dir: Path | None = None) -> Scope:
    home = Path.home()
    if scope == "user":
        return Scope("user", home / ".claude" / "settings.json", "~/.claude/settings.json")
    pd = project_dir or Path.cwd()
    if scope == "project":
        return Scope("project", pd / ".claude" / "settings.json", str(pd / ".claude/settings.json"))
    if scope == "local":
        return Scope("local", pd / ".claude" / "settings.local.json", str(pd / ".claude/settings.local.json"))
    raise ValueError(f"Unknown scope: {scope}")


def _build_hook_entry() -> dict:
    return {
        "matcher": "Skill",
        "hooks": [
            {
                "type": "command",
                "command": str(HOOK_SCRIPT),
                "async": True,
                "timeout": 5,
                # Custom marker — Claude Code ignores unknown keys, but they let us
                # find our entry on uninstall without depending on the path matching.
                "_id": HOOK_MARKER,
            }
        ],
    }


def _read_settings(path: Path) -> dict:
    if not path.exists():
        return {}
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        raise SystemExit(f"refusing to edit malformed settings file {path}: {e}")


def _write_atomic(path: Path, data: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    os.replace(tmp, path)


def _backup(path: Path) -> Path | None:
    """Make a timestamped backup. Also keep the pristine backup if absent."""
    if not path.exists():
        return None
    ts = datetime.now().strftime("%Y%m%d-%H%M%S")
    ts_backup = path.with_name(path.name + f".bak.{ts}")
    shutil.copy2(path, ts_backup)
    # First-time pristine backup (only for the user-scope file)
    if path == Path.home() / ".claude" / "settings.json" and not PRISTINE_BACKUP.exists():
        shutil.copy2(path, PRISTINE_BACKUP)
    return ts_backup


def _has_our_hook(entries: list, command: str) -> bool:
    for matcher_block in entries or []:
        if not isinstance(matcher_block, dict):
            continue
        if matcher_block.get("matcher") != "Skill":
            continue
        for h in matcher_block.get("hooks") or []:
            if isinstance(h, dict) and (h.get("_id") == HOOK_MARKER or h.get("command") == command):
                return True
    return False


def _strip_our_hook(entries: list, command: str) -> list:
    out = []
    for matcher_block in entries or []:
        if not isinstance(matcher_block, dict) or matcher_block.get("matcher") != "Skill":
            out.append(matcher_block)
            continue
        kept = [
            h for h in matcher_block.get("hooks") or []
            if not (isinstance(h, dict) and (h.get("_id") == HOOK_MARKER or h.get("command") == command))
        ]
        if kept:
            new_block = dict(matcher_block)
            new_block["hooks"] = kept
            out.append(new_block)
        # else: drop empty matcher block entirely
    return out


def install(scope: str = "user", project_dir: Path | None = None) -> None:
    if not HOOK_SCRIPT.exists():
        raise SystemExit(f"hook script not found: {HOOK_SCRIPT}")
    if not os.access(HOOK_SCRIPT, os.X_OK):
        # Best-effort: make it executable
        os.chmod(HOOK_SCRIPT, 0o755)

    s = resolve_scope(scope, project_dir)
    settings = _read_settings(s.path)
    hooks = settings.setdefault("hooks", {})
    entry = _build_hook_entry()

    actions = []
    for event in ("PreToolUse", "PostToolUse"):
        existing = hooks.setdefault(event, [])
        if _has_our_hook(existing, str(HOOK_SCRIPT)):
            actions.append(f"  {event}: already present")
            continue
        existing.append(entry)
        actions.append(f"  {event}: added")

    backup = _backup(s.path)
    _write_atomic(s.path, settings)

    print(f"✓ Hooks installed in {s.label} (scope: {s.name})")
    for line in actions:
        print(line)
    if backup:
        print(f"  Backup: {backup}")
    print(f"\n  Hook script: {HOOK_SCRIPT}")
    print(f"  Events log:  ~/.claude/skills-usage/hook-events.jsonl")
    print(f"\nDisable with: skills-usage disable-hook --scope {s.name}")


def uninstall(scope: str = "user", project_dir: Path | None = None) -> None:
    s = resolve_scope(scope, project_dir)
    if not s.path.exists():
        print(f"No settings file at {s.label}; nothing to remove.")
        return

    settings = _read_settings(s.path)
    hooks = settings.get("hooks") or {}
    if not hooks:
        print(f"No hooks block in {s.label}; nothing to remove.")
        return

    changed = False
    for event in ("PreToolUse", "PostToolUse"):
        before = hooks.get(event) or []
        after = _strip_our_hook(before, str(HOOK_SCRIPT))
        if after != before:
            changed = True
            if after:
                hooks[event] = after
            else:
                hooks.pop(event, None)

    if not changed:
        print(f"skills-usage hooks not found in {s.label}; nothing to remove.")
        return

    if not hooks:
        settings.pop("hooks", None)

    backup = _backup(s.path)
    _write_atomic(s.path, settings)
    print(f"✓ Hooks removed from {s.label}")
    if backup:
        print(f"  Backup: {backup}")


def status(scope: str = "user", project_dir: Path | None = None) -> None:
    s = resolve_scope(scope, project_dir)
    if not s.path.exists():
        print(f"{s.label}: file does not exist (no hook installed)")
        return
    settings = _read_settings(s.path)
    hooks = settings.get("hooks") or {}
    pre = _has_our_hook(hooks.get("PreToolUse") or [], str(HOOK_SCRIPT))
    post = _has_our_hook(hooks.get("PostToolUse") or [], str(HOOK_SCRIPT))
    state = "enabled" if (pre and post) else ("partial" if (pre or post) else "disabled")
    print(f"{s.label}: {state}")
    print(f"  PreToolUse:  {'✓' if pre else '·'}")
    print(f"  PostToolUse: {'✓' if post else '·'}")
    print(f"  Hook script: {HOOK_SCRIPT} ({'exists' if HOOK_SCRIPT.exists() else 'MISSING'})")

File: src/hookInstall.ts
import { copyFileSync, existsSync, mkdirSync, readFileSync, renameSync, writeFileSync, chmodSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Bundled hook script that ships with the package (in dist/).
export const HOOK_SCRIPT_BUNDLED = resolve(__dirname, "hookEvent.js");

// Stable, package-manager-independent location the hook command points at.
// We copy the bundled script here so settings.json doesn't reference an
// ephemeral npx cache path (which npm may garbage-collect).
// Resolved via a function so tests can override via SKILLS_USAGE_BIN_DIR.
export function installedScriptPath(): string {
  const override = process.env.SKILLS_USAGE_BIN_DIR;
  const base = override ?? join(homedir(), ".claude", "skills-usage", "bin");
  return join(base, "hookEvent.js");
}

// Backwards-compat alias — eagerly resolved at module load. Most callers should
// prefer installedScriptPath() so tests can override after import.
export const HOOK_SCRIPT_INSTALLED = installedScriptPath();
export const HOOK_SCRIPT = HOOK_SCRIPT_INSTALLED;

export const PRISTINE_BACKUP = join(homedir(), ".claude", "settings.json.pre-skills-usage.bak");
export const HOOK_MARKER = "skills-usage:skill-events";

export type Scope = "user" | "project" | "local";

interface ScopeInfo { name: Scope; path: string; label: string; }

export function resolveScope(scope: Scope, projectDir: string = process.cwd()): ScopeInfo {
  const home = homedir();
  if (scope === "user") {
    return { name: "user", path: join(home, ".claude", "settings.json"), label: "~/.claude/settings.json" };
  }
  if (scope === "project") {
    const p = join(projectDir, ".claude", "settings.json");
    return { name: "project", path: p, label: p };
  }
  if (scope === "local") {
    const p = join(projectDir, ".claude", "settings.local.json");
    return { name: "local", path: p, label: p };
  }
  throw new Error(`Unknown scope: ${scope}`);
}

interface HookCommand { type: string; command: string; async?: boolean; timeout?: number; _id?: string; [k: string]: unknown; }
interface MatcherBlock { matcher: string; hooks: HookCommand[]; }
interface HooksConfig { PreToolUse?: MatcherBlock[]; PostToolUse?: MatcherBlock[]; [k: string]: MatcherBlock[] | undefined; }

function buildEntry(): MatcherBlock {
  // Use absolute path to the installed Node script. Node hooks need to be invoked
  // by node — chmod +x on a .js file isn't enough on every system, so we run via node.
  const command = `node ${installedScriptPath()}`;
  return {
    matcher: "Skill",
    hooks: [
      {
        type: "command",
        command,
        async: true,
        timeout: 5,
        _id: HOOK_MARKER,
      },
    ],
  };
}

function readSettings(path: string): Record<string, any> {
  if (!existsSync(path)) return {};
  try {
    return JSON.parse(readFileSync(path, "utf-8"));
  } catch (e) {
    throw new Error(`refusing to edit malformed settings file ${path}: ${e}`);
  }
}

function writeAtomic(path: string, data: Record<string, any>): void {
  mkdirSync(dirname(path), { recursive: true });
  const tmp = path + ".tmp";
  writeFileSync(tmp, JSON.stringify(data, null, 2) + "\n", "utf-8");
  renameSync(tmp, path);
}

function backup(path: string): string | null {
  if (!existsSync(path)) return null;
  const d = new Date();
  const pad2 = (n: number) => String(n).padStart(2, "0");
  const ts =
    `${d.getFullYear()}${pad2(d.getMonth() + 1)}${pad2(d.getDate())}-` +
    `${pad2(d.getHours())}${pad2(d.getMinutes())}${pad2(d.getSeconds())}`;
  const tsBackup = `${path}.bak.${ts}`;
  copyFileSync(path, tsBackup);
  if (path === join(homedir(), ".claude", "settings.json") && !existsSync(PRISTINE_BACKUP)) {
    copyFileSync(path, PRISTINE_BACKUP);
  }
  return tsBackup;
}

function hasOurHook(entries: MatcherBlock[] | undefined, command: string): boolean {
  if (!entries) return false;
  for (const block of entries) {
    if (block?.matcher !== "Skill") continue;
    for (const h of block.hooks ?? []) {
      if (h?._id === HOOK_MARKER) return true;
      if (h?.command === command) return true;
    }
  }
  return false;
}

function stripOurHook(entries: MatcherBlock[] | undefined, command: string): MatcherBlock[] {
  if (!entries) return [];
  const out: MatcherBlock[] = [];
  for (const block of entries) {
    if (block?.matcher !== "Skill") {
      out.push(block);
      continue;
    }
    const kept = (block.hooks ?? []).filter((h) => h?._id !== HOOK_MARKER && h?.command !== command);
    if (kept.length) out.push({ ...block, hooks: kept });
  }
  return out;
}

export function install(scope: Scope = "user", projectDir?: string): void {
  if (!existsSync(HOOK_SCRIPT_BUNDLED)) {
    throw new Error(`bundled hook script not found: ${HOOK_SCRIPT_BUNDLED}\n(did the package's dist/ install correctly?)`);
  }

  // Copy the bundled hookEvent.js to a stable per-user location so the
  // settings.json command stays valid even if this package was launched
  // from an ephemeral npx cache that gets garbage-collected later.
  const installed = installedScriptPath();
  mkdirSync(dirname(installed), { recursive: true });
  copyFileSync(HOOK_SCRIPT_BUNDLED, installed);
  try { chmodSync(installed, 0o755); } catch {}

  const s = resolveScope(scope, projectDir);
  const settings = readSettings(s.path);
  if (!settings.hooks) settings.hooks = {};
  const hooks = settings.hooks as HooksConfig;
  const entry = buildEntry();
  const cmd = entry.hooks[0].command;

  const actions: string[] = [];
  for (const event of ["PreToolUse", "PostToolUse"] as const) {
    if (!hooks[event]) hooks[event] = [];
    if (hasOurHook(hooks[event], cmd)) {
      actions.push(`  ${event}: already present`);
      continue;
    }
    hooks[event]!.push(entry);
    actions.push(`  ${event}: added`);
  }

  const bk = backup(s.path);
  writeAtomic(s.path, settings);

  console.log(`✓ Hooks installed in ${s.label} (scope: ${s.name})`);
  for (const line of actions) console.log(line);
  if (bk) console.log(`  Backup: ${bk}`);
  console.log(`\n  Hook script: ${installed}`);
  console.log(`  Events log:  ~/.claude/skills-usage/hook-events.jsonl`);
  console.log(`\nDisable with: skills-usage disable-hook --scope ${s.name}`);
}

export function uninstall(scope: Scope = "user", projectDir?: string): void {
  const s = resolveScope(scope, projectDir);
  if (!existsSync(s.path)) {
    console.log(`No settings file at ${s.label}; nothing to remove.`);
    return;
  }
  const settings = readSettings(s.path);
  const hooks = settings.hooks as HooksConfig | undefined;
  if (!hooks) {
    console.log(`No hooks block in ${s.label}; nothing to remove.`);
    return;
  }

  const cmd = `node ${installedScriptPath()}`;
  let changed = false;
  for (const event of ["PreToolUse", "PostToolUse"] as const) {
    const before = hooks[event] ?? [];
    const after = stripOurHook(before, cmd);
    if (JSON.stringify(before) !== JSON.stringify(after)) {
      changed = true;
      if (after.length) hooks[event] = after;
      else delete hooks[event];
    }
  }
  if (!changed) {
    console.log(`skills-usage hooks not found in ${s.label}; nothing to remove.`);
    return;
  }
  if (Object.keys(hooks).length === 0) delete settings.hooks;

  const bk = backup(s.path);
  writeAtomic(s.path, settings);
  console.log(`✓ Hooks removed from ${s.label}`);
  if (bk) console.log(`  Backup: ${bk}`);
}

export function status(scope: Scope = "user", projectDir?: string): void {
  const s = resolveScope(scope, projectDir);
  if (!existsSync(s.path)) {
    console.log(`${s.label}: file does not exist (no hook installed)`);
    return;
  }
  const settings = readSettings(s.path);
  const hooks = (settings.hooks ?? {}) as HooksConfig;
  const installed = installedScriptPath();
  const cmd = `node ${installed}`;
  const pre = hasOurHook(hooks.PreToolUse, cmd);
  const post = hasOurHook(hooks.PostToolUse, cmd);
  const state = pre && post ? "enabled" : pre || post ? "partial" : "disabled";
  console.log(`${s.label}: ${state}`);
  console.log(`  PreToolUse:  ${pre ? "✓" : "·"}`);
  console.log(`  PostToolUse: ${post ? "✓" : "·"}`);
  console.log(`  Hook script: ${installed} (${existsSync(installed) ? "exists" : "MISSING"})`);
}
